### PR TITLE
SPDX expression support improvements

### PR DESCRIPTION
**Pull request by nscuro**
### Description

Building upon #2400, this PR adds missing functionality for setting license expressions via REST API, and adds expression validation for both REST API and BOM ingestion.

Invalid expressions submitted via frontend or REST API will result in an API error, whereas invalid expressions encountered during BOM ingestion will be logged and skipped.

### Addressed Issue

N/A

### Additional Details

N/A

### Checklist

- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- ~This PR fixes a defect, and I have provided tests to verify that the fix is effective~
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- ~This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)~
- ~This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly~

### CODE REVIEW

**Comment by hborchardt** [comment]
Thanks for following up on this, and making improvements where I was lacking knowledge about the framework's capabilities or intricacies 🚀

**Comment by melba-lopez** [comment]
> Thanks for following up on this, and making improvements where I was lacking knowledge about the framework's capabilities or intricacies 🚀

Thank you @hborchardt for contributing what you have been able to!  Sometimes we don't get enough time or have the knowledge _(im sure you can reference my PR thats been open for eons now 😸 )_

File: src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java [hunk 1]
```diff
@@ -136,7 +130,7 @@ public void informTest() throws Exception {
         final byte[] bomBytes = Files.readAllBytes(Paths.get(getClass().getClassLoader().getResource("bom-1.xml").toURI()));
 
         new BomUploadProcessingTask().inform(new BomUploadEvent(project.getUuid(), bomBytes));
-        assertConditionWithTimeout(() -> NOTIFICATIONS.size() >= 5, Duration.ofSeconds(5));
+        assertConditionWithTimeout(() -> NOTIFICATIONS.size() >= 6, Duration.ofSeconds(5));
```
Comment (melba-lopez) [inline src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:133]: i see this is being changed from notification size `>=  5` ... to `6`.  Is this a 1 to 1 mapping with duration of seconds?   @nscuro

File: src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java [hunk 2]
```diff
@@ -221,4 +215,127 @@ public void informIssue2859Test() throws Exception {
...
+        license.setName("Eclipse Public License 2.0");
+        qm.persist(license);
+
+        final var project = new Project();
+        project.setName("acme-app");
+        qm.persist(project);
+
+        final byte[] bomBytes = """
+                {
+                  "bomFormat": "CycloneDX",
```
Comment (melba-lopez) [inline src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:270]: if this is a test to upload a bom that is of SPDX, should we include appropriate spdx nomenclature outside of just the license?  I suspect if someone uploads an SPDX bom, they would want that information to stay when downloading it too.

**Review by melba-lopez — COMMENTED** [review]
@nscuro 👨‍💻 guru! awesome changes in such a short amount of time.  I had 2 comments that could be easy for you to answer/address.  Hopefully you can take a look before merging!

File: src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java (hunk 1 above)
Comment (nscuro) [reply src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:133]: It's an inconsistency in assertions that I noticed while debugging another issue with this test. 

This statement is waiting for 6 (previously 5) notifications to be sent, and will wait up to 5 seconds for this to happen.
It's more or less a means of ensuring all asynchronous processing has completed before doing further assertions.

The actual notification content is being asserted later in the test:

https://github.com/DependencyTrack/dependency-track/blob/06168ebaa38a5b027c087e5d2d41b98aa61882a2/src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java#L155-L170

I noticed that the above assertion is checking for **6** notifications, not just 5... :D

File: src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java (hunk 2 above)
Comment (nscuro) [reply src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:270]: SPDX BOMs are not supported at all right now; Uploads of such BOMs will be rejected. We have https://github.com/DependencyTrack/dependency-track/issues/1746 open, but doesn't look like we're going to address that anytime soon.

For this change, we just want to ensure that SPDX license expressions within a CDX BOM are correctly ingested.

Comment (melba-lopez) [reply src/test/java/org/dependencytrack/tasks/BomUploadProcessingTaskTest.java:270]: 🤦‍♀️ for various reasons we discussed... i swear it was!  Thanks for bringing me back to reality 😺

**Pull request merged**